codalab/worker_manager/worker_manager.py

```diff
@@ -0,0 +1,167 @@
+import logging
+import time
+from codalab.lib.codalab_manager import CodaLabManager
+from codalab.worker.bundle_state import State
+from collections import namedtuple
+
+logger = logging.getLogger(__name__)
+
+# Represents a AWS/Azure job that runs a single cl-worker.
+# For now, we don't need to keep any information.
+WorkerJob = namedtuple('WorkerJob', [])
+
+
+class WorkerManager(object):
+    """
+    The abstract class for a worker manager.  Different backends like AWS,
+    Azure, Slurm should override `get_worker_jobs` and `start_worker_job`.
+
+    The basic architecture of the WorkerManager is extremely simple: to a
+    first-order approximation, it simply launches `cl-worker`s as AWS/Azure
+    batch jobs as long as there staged bundles.
+
+    The simplicity means that we don't need to manage state about how workers
```

Comment:
Architecturally, I'm wondering whether it would be better to have the worker manager handle this complexity / state instead of `cl-worker`. I'm not sure about the benefits or drawbacks of either approach, though -- do you have any thoughts?

Reply:
We have tried putting things into `cl-worker` and it has always turned into a nightmare.

Currently the implementation is pretty hands off (doesn't associate workers with jobs), but we could also attach particular bundle uuids to workers when we launch them, so I think we're not really losing on generality.